File: src/main/java/edu/cornell/cs/apl/viaduct/util/dataflow/DataFlowEdge.java
package edu.cornell.cs.apl.viaduct.util.dataflow;

/**
 * Edges of a data flow graph. These edges are allowed to modify the values passing through them.
 */
public interface DataFlowEdge<A> {
  /**
   * Apply this function to the value incoming from the source node before passing it onto the
   * destination node.
   */
  A propagate(A in);
}

File: compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/util/dataflow/DataFlowEdge.kt
package edu.cornell.cs.apl.viaduct.util.dataflow

/** Edges in a data flow graph. These are allowed to modify values as they pass through them. */
interface DataFlowEdge<A> {
    /**
     * Apply this function to the value incoming from the source node before passing it onto the
     * destination node.
     */
    fun propagate(input: A): A
}
